src/java_bytecode/jar_file.cpp

```diff
@@ -1,131 +1,86 @@
+  m_zip_archive(filename)
 {
-  if(!mz_ok)
+  const size_t file_count=m_zip_archive.get_num_files();
+  for(size_t file_id=0; file_id<file_count; file_id++)
   {
-    memset(&zip, 0, sizeof(zip));
-    mz_bool mz_open=mz_zip_reader_init_file(&zip, filename.c_str(), 0);
-    mz_ok=mz_open==MZ_TRUE;
-  }
-
-  if(mz_ok)
-  {
-    std::size_t number_of_files=
-      mz_zip_reader_get_num_files(&zip);
-
-    for(std::size_t i=0; i<number_of_files; i++)
-    {
-      mz_uint filename_length=mz_zip_reader_get_filename(&zip, i, nullptr, 0);
-      char *filename_char=new char[filename_length+1];
-      mz_uint filename_len=
-        mz_zip_reader_get_filename(&zip, i, filename_char, filename_length);
-      assert(filename_length==filename_len);
-      std::string file_name(filename_char);
-
-      // non-class files are loaded in any case
-      bool add_file=!has_suffix(file_name, ".class");
-      // load .class file only if they match regex / are in match set
-      add_file|=class_loader_limit.load_class_file(file_name);
-      if(add_file)
-      {
-        if(has_suffix(file_name, ".class"))
-          status() << "read class file " << file_name
-                   << " from " << filename << eom;
-        filtered_jar[file_name]=i;
-      }
-      delete[] filename_char;
-    }
+    const auto filename=m_zip_archive.get_filename(file_id);
+    if(!has_suffix(filename, ".class") || limit.load_class_file(filename))
+      m_name_to_id.emplace(filename, file_id);
   }
 }
 
-jar_filet::~jar_filet()
+std::string jar_filet::get_entry(const std::string &name)
 {
-  if(mz_ok)
+  const auto entry=m_name_to_id.find(name);
+  INVARIANT(entry!=m_name_to_id.end(), "File doesn't exist");
+  try
+  {
+    return m_zip_archive.extract(entry->second);
+  }
+  catch(const std::runtime_error &)
   {
-    mz_zip_reader_end(&zip);
-    mz_ok=false;
+    return "";
   }
 }
 
-std::string jar_filet::get_entry(const irep_idt &name)
+/// Remove leading and trailing whitespace characters from string
+static std::string trim(
+  std::string::const_iterator begin,
+  std::string::const_iterator end)
 {
-  if(!mz_ok)
-    return std::string("");
-
-  std::string dest;
-
-  auto entry=filtered_jar.find(name);
-  assert(entry!=filtered_jar.end());
-
-  size_t real_index=entry->second;
-  mz_zip_archive_file_stat file_stat;
-  memset(&file_stat, 0, sizeof(file_stat));
-  mz_bool stat_ok=mz_zip_reader_file_stat(&zip, real_index, &file_stat);
-  if(stat_ok!=MZ_TRUE)
-    return std::string();
-  std::vector<char> buffer;
-  size_t bufsize=file_stat.m_uncomp_size;
-  buffer.resize(bufsize);
-  mz_bool read_ok=
-    mz_zip_reader_extract_to_mem(&zip, real_index, buffer.data(), bufsize, 0);
-  if(read_ok!=MZ_TRUE)
-    return std::string();
-
-  dest.insert(dest.end(), buffer.begin(), buffer.end());
-
-  return dest;
+  while(begin!=end && std::isspace(*begin))
+    ++begin;
+  if(begin!=end)
+  {
+    --end;
+    while(begin!=end && std::isspace(*end))
+      --end;
+    end++;
```

Comment:
AFAIK this is just for the `make_...` convenience function, and `reverse_iterator` itself should still be usable in C++98 onwards